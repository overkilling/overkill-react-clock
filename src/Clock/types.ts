export interface Time {
  time: Date
}
