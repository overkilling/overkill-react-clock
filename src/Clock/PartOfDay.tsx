import * as React from 'react'
import { Time } from './types'
import './Clock.css'
import './PartOfDay.css'

const partAt = (time: Date) => {
  const hour = time.getHours()

  if (hour >= 5 && hour < 12) {
    return 'morning'
  } else if (hour >= 12 && hour < 17) {
    return 'afternoon'
  } else if (hour >= 17 && hour < 21) {
    return 'evening'
  }
  return 'night'
}

export const PartOfDay = ({ time }: Time) => (
  <p className="partOfDay">It is {partAt(time)}</p>
)
