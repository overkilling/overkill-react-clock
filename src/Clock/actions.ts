export type UpdateClock = 'UPDATE_CLOCK'

export interface UpdateClockAction {
  type: UpdateClock
  time: Date
}

export const updateClock = (newTime: Date): UpdateClockAction => ({
  type: 'UPDATE_CLOCK',
  time: newTime
})
