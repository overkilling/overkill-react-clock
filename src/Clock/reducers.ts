import type { UpdateClockAction } from './actions'
import { Time } from './types'

const initialState = {
  time: new Date()
}

export const timeKeeper = (
  state: Time = initialState,
  { type, time }: UpdateClockAction
): Time => {
  switch (type) {
    case 'UPDATE_CLOCK':
      return { time }
    default:
      return state
  }
}

export default timeKeeper
