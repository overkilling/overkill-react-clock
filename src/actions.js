// @flow

export type UpdateClock = 'UPDATE_CLOCK';

export type UpdateClockAction = {
  type: UpdateClock,
  time: Date
};

export const updateClock = (newTime: Date): UpdateClockAction => ({
    type: 'UPDATE_CLOCK',
    time: newTime,
});
