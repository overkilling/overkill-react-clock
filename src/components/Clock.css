.clock {
    font-family: "Orbitron";
    color: rgb(40, 182, 27);
    font-size: 200%;
}
