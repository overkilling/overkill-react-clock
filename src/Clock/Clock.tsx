import * as React from 'react'
import { useSelector } from 'react-redux'
import { PartOfDay } from './PartOfDay'
import { TimeDisplay } from './TimeDisplay'
import { CombinedState } from '../types'

export const Clock = () => {
  const time = useSelector((state: CombinedState) => state.clock.time)
  const { format, timezone } = useSelector(
    (state: CombinedState) => state.configuration
  )

  return (
    <div>
      <TimeDisplay time={time} format={format} timezone={timezone} />
      <PartOfDay time={time} />
    </div>
  )
}
