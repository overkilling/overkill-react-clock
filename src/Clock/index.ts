import { TimeDisplay } from './TimeDisplay/TimeDisplay'
import { PartOfDay } from './PartOfDay/PartOfDay'

export { TimeDisplay, PartOfDay }
