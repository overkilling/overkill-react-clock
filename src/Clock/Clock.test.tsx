import * as React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { createStore, combineReducers } from 'redux'
import { Clock } from './Clock'
import reducer from './reducers'
import configurationReducer from '../Configuration/reducers'
import { updateClock } from './actions'

describe('Clock component', () => {
  it('renders clock', () => {
    const date = new Date(2018, 2, 11, 11, 49, 13)
    const testStore = createStore(
      combineReducers({
        clock: reducer,
        configuration: configurationReducer
      }),
      {
        clock: { time: date },
        configuration: { format: 'HH:mm:ss', timezone: 'America/New_York' }
      }
    )
    const { asFragment } = render(
      <Provider store={testStore}>
        <Clock />
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders on UPDATE_CLOCK action', () => {
    const testStore = createStore(
      combineReducers({
        clock: reducer,
        configuration: configurationReducer
      }),
      { configuration: { format: 'h:mm:ss A', timezone: 'Zulu' } }
    )
    const { getByText } = render(
      <Provider store={testStore}>
        <Clock />
      </Provider>
    )
    const newDate = new Date(2018, 2, 11, 16, 30, 7)

    testStore.dispatch(updateClock(newDate))

    expect(getByText('4:30:07 PM')).toBeInTheDocument()
    expect(getByText('It is afternoon')).toBeInTheDocument()
  })
})
