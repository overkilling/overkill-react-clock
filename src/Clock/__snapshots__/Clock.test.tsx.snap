// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clock component renders clock 1`] = `
<DocumentFragment>
  <div>
    <p
      class="clock"
    >
      07:49:13
    </p>
    <p
      class="partOfDay"
    >
      It is morning
    </p>
  </div>
</DocumentFragment>
`;
